### PR TITLE
feat: implement `+loading.svelte`

### LINKED ISSUE #7213: option to render layout before page is loaded
### Describe the problem

When building application-like websites, users expect instant feedback when changing pages, even if the page's content isn't ready yet. It's confusing to a user when they click a button and nothing happens, then a second later the page changes. Loading indicators help but are still not ideal.

### Describe the proposed solution

Provide an option that will allow a `+layout.svelte` file to be rendered before the `+page.svelte` is ready. For example, say clicking a button opens up a page that is a modal. We place the markup for the modal inside the `+layout.svelte` file and the content inside the `+page.svelte` file. As soon as the user clicks, we can render an empty modal. Then once the page is ready, the content will appear. While the page is loading, we can utilize the slot default as a loading page or skeleton UI.
```html
<script>
	// declare that this layout will be rendered to the DOM before the page is loaded
	export const optimistic = true;
</script>

<div class="modal">
	<slot>
		<p>skeleton ui or loading indicator</p>
	</slot>
</div>
```

### Alternatives considered

- **Using `#await` block so that surrounding markup is not dependent on data:** This destroys SSR and sveltekit's entire data model.
- **Having a dedicated `+loading.svelte` file:** Seems like unnecessary complexity when it can be done without it. If for some reason someone requires separate default and placeholder page contents they can always use the `$navigating` store to switch between them.

### Importance

would make my life easier

### Additional Information

- If the `+layout.svelte` also relies on data we can still render it as soon as its own data is ready.
- This probably shouldn't affect SSR in any way. When we render on the server, we probably want to wait until the entire page is loaded before responding to the client. If we didn't, it would defeat the purpose of SSR in the first place.
- As far as I know, this doesn't break any existing functionality.

### PR DESCRIPTION
**Pull request by dummdidumm**
**Punting on this for now, the related issue brought up more and more questions that hint at that this isn't the right approach, and something like `defer` and using `{#await ..}` is more promising. There's also unsolved obstacles around server data (which we can't stream, because of cookies [they could be set by the user after the response has started streaming, at which point it's too late for headers to be set]).**

closes #7213

Super rough draft for now.

Things I noticed:
- `client.js` really is becoming harder and harder to reason about. We need to rework this, I'll try to come up with something. I spent far too much time reading just the `load_route` function over and over again to understand what happens when and how and in what order. Sure, some of it we won't solve with code because it's just complicated, but we can do better
- `PageNode` is a "this can contains this this this all optional" when in reality it's "can contain this OR this". This makes it harder to reason about from a types perspective, although it makes some code easier to write probably. maybe something for a later refactoring

Things to take care of:
- [x] preloading should probably also invoke the `+loading.svelte` logic, but if half of the stuff is loaded, and then the user clicks on the link, we need to know at which point in the loading chain we are to load the right `+loading.svelte` (or none)
- if we have configurable parameters, we also need to pass them to the frontend like we do with `trailingSlash`

Questions that came up for me:
- should multiple `+loading.svelte` between two layouts or a layout and a page be an error? Probably. Do we check this for `+error.svelte` already? Probably not. Example:
```
routes
  +layout
  path
    +loading
     topage
        +loading
        +page
``` 
- Is a lone `+loading.svelte` (not next to a layout or page) an error? Maybe? I don't think we check the same for `+error.svelte` yet, and if we started, it's probably seen as a breaking change? Example:
```
routes
  +layout
  path
    +loading
     topage
        +page
``` 
- how does `+loading.svelte` interact with scroll handling and focus management? Since we agreed that the URL is updated eagerly, so should scroll handling and focus management in my opinion. I'm not sure what this means if an element with autofocus is shown later, I'd hope the browser handles this automatically.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: documentation/docs/30-advanced/10-advanced-routing.md [hunk 1]
```diff
@@ -2,6 +2,38 @@
 title: Advanced routing
 ---
 
+## +loading.svelte
```
Comment (dummdidumm) [inline documentation/docs/30-advanced/10-advanced-routing.md:5]: Wasn't really sure if this belongs to "core concepts" or "advanced" - I put it in the latter

File: documentation/docs/30-advanced/10-advanced-routing.md [hunk 2]
```diff
@@ -2,6 +2,38 @@
...
+
+Not all data is available fast enough to provide a pleasant navigation experience. `+loading.svelte` helps you showing meaningfull fallback UI while data for a page is still loading. Consider the following folder structure:
```
Comment (tomecko) [inline documentation/docs/30-advanced/10-advanced-routing.md:7]:
```suggestion
Not all data is available fast enough to provide a pleasant navigation experience. `+loading.svelte` helps you showing meaningful fallback UI while data for a page is still loading. Consider the following folder structure:
```

File: packages/kit/src/runtime/client/client.js [hunk 3]
```diff
@@ -490,6 +481,53 @@ export function create_client({ target, base }) {
...
+	 */
+	function normalize_url(branch, url) {
+		/** @type {import('types').TrailingSlash} */
+		let slash = 'never';
+		for (const node of branch) {
+			if (node?.slash !== undefined) slash = node.slash;
+		}
+		// TODO we manipulate the incoming URL here, check what implications this has
+		url.pathname = normalize_path(url.pathname, slash);
+		url.search = url.search; // turn `/?` into `/`
```
Comment (tomecko) [inline packages/kit/src/runtime/client/client.js:496]: does it have any effect?

File: packages/kit/test/apps/basics/test/client.test.js [hunk 4]
```diff
@@ -1254,3 +1254,45 @@ test.describe('Interactivity', () => {
...
+
+test.describe('+loading.svelte', () => {
+	test('shows loading ui', async ({ page, clicknav }) => {
+		await page.goto('/loading-ui');
+		clicknav('.client-load-1'); // do not await!
+		await page.waitForTimeout(100); // give time to rerender
+		expect(await page.textContent('h3')).toBe('Loading');
+		await page.waitForTimeout(1200);
+		expect(await page.textContent('h3')).toBe('Result');
+	});
```
Comment (tomecko) [inline packages/kit/test/apps/basics/test/client.test.js:1266]: I think `waitForTimeout` are not necessary here and can be avoided. `timeout` option can be passed to `textContent` if needed.

File: packages/kit/test/apps/basics/test/client.test.js [hunk 5]
```diff
@@ -1254,3 +1254,45 @@ test.describe('Interactivity', () => {
...
+
+	test('skips loading ui (data-sveltekit-preload-data)', async ({ page, clicknav }) => {
+		await page.goto('/loading-ui');
+		await page.hover('.client-load-2');
+		await page.waitForTimeout(1500);
+		clicknav('.client-load-2'); // do not await!
+		await page.waitForTimeout(100); // give time to rerender
+		expect(await page.textContent('h3')).toBe('Result');
```
Comment (tomecko) [inline packages/kit/test/apps/basics/test/client.test.js:1296]: would it be reasonable to assert of non-existence of "Loading" here?

**Comment by Rich-Harris** [comment]
Closing as we decided not to do this, we'll follow up with a different approach in the near future

**Pull request closed without merging**